### DISCUSSION TITLE
message: "Invalid stereotool library"

### DISCUSSION THREAD

**Discussion by PetaVolt** [savonet/liquidsoap#3448] [User Support]
Hello,

I've recently upgraded liquidsoap to version 2.2.1 with intentions to start using stereotool. I'm running on Ubuntu 22.04 x64. Following the docs, I adding the stereotool operator in the script and it results in this error:

```
stream liquidsoap[8564]: Error 14: Uncaught runtime error:
stream liquidsoap[8564]: type: invalid,
stream liquidsoap[8564]: message: "Invalid stereotool library",
stream liquidsoap[8564]: stack: at audio.liq, line 232, char 6-73,
```
Here's a snippet of the stereotool block that's being used:

```
final = stereotool(
  library_file="/home/spencer/Stereo_Tool_Generic_plugin_992",
  license_key="<3f...2a>",
  preset="/home/spencer/audio-files/liquidsoap/preset.sts",
  final
)
```

Commenting this block of code results in no errors and works without issue. It's also worth noting I've tried stereotool versions 10.10 and 9.92, same results on both. Help would be appreciated.

Thanks!
Spencer

**Comment by vitoyucepi** [comment]
Hi @PetaVolt,
The Stereo Tool plugin has a strange packaging. It provides an all-in-one archive of plugins for all operating systems and all architectures.

Could you please describe step by step how you downloaded and installed the plugin.
1. Go to https://www.thimeo.com/stereo-tool/download-992/.
1. Scroll down to *x86/64 Linux Thimeo-ST Plugin*.
1. Right-click the *Plugin SDK with binaries for all OS's* button.
1. Select *Copy link address* from the context menu.
1. On the server, use the commands
   ```sh
   wget https://download.thimeo.com/Stereo_Tool_Generic_plugin_992.zip
   unzip Stereo_Tool_Generic_plugin_992.zip
   ```
1. In liquidsoap specify
   ```ruby
   final = stereotool(
     library_file="/home/spencer/Stereo_Tool_Generic_plugin_992/libStereoTool_64.so",
     license_key="<3f...2a>",
     preset="/home/spencer/audio-files/liquidsoap/preset.sts",
     final
   )
   ```

**Comment by PetaVolt** [comment]
Hello Vito,

Thanks for the explanation. I updated the script as you specified and that did the trick. My understanding was liquidsoap only needed the directory, rather than the file.

I'm now running into a separate issue where the script fails when running in --daemon mode but works fine otherwise.  I'll continue to dig into this.

**Reply by vitoyucepi** [reply]
If you are planning to run liquidsoap with systemd, then I can recommend that you take a look at #2558 and #3354.

**Comment by mcfiredrill** [comment]
I've been using the stereotool operator for a long time, suddenly it's not working for me anymore and I get this error. I am sure I have the right library and its readable. I've tried downgrading to previous version of liquidsoap.

**Reply by mcfiredrill** [reply]
It seems to work fine on my linux machine, but not in my Docker container build.

Don't really know what I'm doing, but I thought I'd make a small ocaml program to try to reproduce how liquidsoap opens the shared library.

```
open Ctypes
open Foreign

let dlopen = foreign "dlopen" (string @-> int @-> returning (ptr void))
let dlerror = foreign "dlerror" (void @-> returning string)

let () =
  let filename = "./libStereoTool_64.so" in
  let rtld_now = 2 in (* RTLD_NOW flag *)
  let handle = dlopen filename rtld_now in
  if handle = null then
    Printf.eprintf "dlopen failed: %s\n" (dlerror ())
  else
    Printf.printf "Library %s loaded successfully!\n" filename

```

```
ocamlfind ocamlopt -o dlopen_test -thread -package ctypes.foreign -linkpkg dlopen_test.ml
```

in my desktop environment:
➜  liquidsoap git:(disconnect) ✗ ./dlopen_test
Library ./libStereoTool_64.so loaded successfully!

in my docker environment:

[liquidsoap@59aba00d06b3 radio]$ ./dlopen_test
dlopen failed: ./libStereoTool_64.so: cannot enable executable stack as shared object requires: Invalid argument


Did some research and asking chat gpt...looks like the library can be modified like this:

execstack -c ./libStereoTool_64.so

Liquidsoap runs with stereotool now!

**Reply by toots** [reply]
That's wild. Thanks for sharing. We should report that to the stereotool upstream.

**Reply by vitoyucepi** [reply]
Hi @mcfiredrill,
What version of glibc are you using? Could you open a new issue or discussion so we can discuss your setup?
